### ISSUE TITLE
When LOQ or non conformity decision is made lock other reviews

### ISSUE THREAD

**Issue by nmadruga** [openmsupply/application-manager-server#364]

**Comment by andreievg** [comment]
https://app.diagrams.net/#G1VIyFc7NjAEShXOQxgJoOvKgKrxbtsY0q

As per above diagram, should lock reviews when consolidation is submitted for anything other then changes requested

**Comment by nmadruga** [comment]
@andreievg I think it would be the case for a consolidation submitted as Changes requested as well. Imagine this case:
1. Reviewer 1 has a draft partial review of Application A
2. Reviewer 2 SUBMIT a partial review of Application A
3. Consolidator SUBMIT back to Reviewer2 with CHANGES REQUESTED
4. Reviewer1 should have their review LOCKED to avoid SUBMITTING and changing Consolidator back to PENDING

If we allow Reviewer1 to SUBMIT while Reviewer2 has CHANGES_REQUESTED status, how would this show to each user?

**Comment by nmadruga** [comment]
Let's try to describe the possible cases here (this will also be available on updated docs I'm planning to create a PR for today BTW - issue #726), when a review is LOCKED/UNLOCKED: 
1. Locking reviews after CHANGES_REQUESTED to Reviewer:
   - Update status to LOCKED for reviews is in DRAFT (lower level to consolidation submitted)
   - Update status to DRAFT for reviews in same level on a review re-submission after CHANGES_REQUEST
2. Locking reviews after LOQ or NO_CONFORM decision to Applicant:
   - Update status to LOCKED for reviews is in DRAFT (lower level to consolidation submitted or same-stage if we consider more than one reviewer with `is-last-level` on that stage)
   - Update status to DRAFT for reviews in first level on a application re-submission after CHANGES_REQUIRED
Q: Locking start of review with assignment as ASSIGNED when CHANGES_REQUESTED to reviewers or when consolidation submits a LOQ or NO_CONFORM to applicant.

**Comment by andreievg** [comment]
As far as I can see its, it's fine to have concurrent change requests to reviewers.

**Comment by andreievg** [comment]
So in my view the logic is this (I think the diagram tried to convey it, but maybe it's too complex):

When consolidator submits LOQ or non conform. Lock all previous level reviews that are draft.

When applicant submits an application we unlock locked ones (which is already implemented)

**Comment by andreievg** [comment]
What to do for reviews that are not started, but LOQ or non conform was submitted ?

If consolidation was only one level we can add extra check to see if application is 'submitted'.

We can have `isLocked` on reviewAssignment, rather than locked status on review. Would need to do a little bit of juggling, but should work for that use case.

**Comment by nmadruga** [comment]
> We can have `isLocked` on reviewAssignment, rather than locked status on review. Would need to do a little bit of juggling, but should work for that use case.

I'm trying to see if I understood this correctly before starting implementing, would the `isLocked` replace the `review.status` "LOCKED" as well? Or would it be a second thing to do once a review is submitted to applicant? In case it's just related to `reviewAssignment` without a review, wouldn't be better to add as another status: "LOCKED"?

Another question - Do you think it's better to create another action just to deal with locking/unlocking reviews on review and application submissions, or include the logic (as already started) as part of the action `updateReviewStatuses`?

**Comment by andreievg** [comment]
> I'm trying to see if I understood this correctly before starting implementing, would the isLocked replace the review.status "LOCKED" as well? Or would it be a second thing to do once a review is submitted to applicant? In case it's just related to reviewAssignment without a review, wouldn't be better to add as another status: "LOCKED"?

Just thought of use case 3, which suggests it should be outside of reviewAssignment status (at least with the current implementation of status). Most times we would know that 'LOCKED' status must have replaced 'ASSIGNED' status, but in the use case where LOQ or non conformity is being submitted when there is still one un-assigned reviewer, we wouldn't know.

Happy to keep 'LOCKED' status on review, but when it's first created, we make sure to check reviewAssignment and mark review as 'LOCKED' if it's indicated as such on reviewAssignment.

Where on reviewAssignment should we store the 'LOCKED' status ? As per above, it's likely not going to work if we store it in the same place that indicates Assigned/Not Assigned yet. Would there be more use cases, other then not allowing not yet created review being submitted, that we may need to consider (i.e. marking some 'status' for review when review is not created yet)

Maybe there is a better solution ?

**Comment by andreievg** [comment]
> Another question - Do you think it's better to create another action just to deal with locking/unlocking reviews on review and application submissions, or include the logic (as already started) as part of the action updateReviewStatuses?

Hmm what do you think ?

**Comment by andreievg** [comment]
Just to sum up, the brainstorm above would require:

Option 1:

* a way to set review to status 'LOCKED' when it's created if reviewAssignment is 'Locked'
* a way to set all review assignments to 'LOCKED', that have no review that are assigned or can be assigned for level -1, when current level submits anything apart form changes requested

Option 2:

* a way to set reviewAssignment to 'LOCKED" for ALL level -1 when current level submits anything apart from changes requested
* change to use 'LOCKED' status from reviewAssignment (can we see updated review assignment status, currently, in the review ?)

**Comment by nmadruga** [comment]
> > I'm trying to see if I understood this correctly before starting implementing, would the isLocked replace the review.status "LOCKED" as well? Or would it be a second thing to do once a review is submitted to applicant? In case it's just related to reviewAssignment without a review, wouldn't be better to add as another status: "LOCKED"?
> 
> Just thought of use case 3, which suggests it should be outside of reviewAssignment status (at least with the current implementation of status). Most times we would know that 'LOCKED' status must have replaced 'ASSIGNED' status, but in the use case where LOQ or non conformity is being submitted when there is still one un-assigned reviewer, we wouldn't know.

Yes, if we do end up adding another status on review_assignment as "TEMPORARY_LOCKED" for example, it would need to be considered in queries for Assigned. I agree that it would break the current logic, so maybe better to keep on a separated field... I'm just concerned how to manage this new flag `isLocked`.

> Happy to keep 'LOCKED' status on review, but when it's first created, we make sure to check reviewAssignment and mark review as 'LOCKED' if it's indicated as such on reviewAssignment.

I see, so we can allow the user to start the review, but not allow to submit the review, I like that idea. Will create a place_holder note for another issue in case we implement using this logic.
